File: lintcode/golang/0969_longest_repeating_substring.go
//hash, binary search

/**
 * @param str: The input string
 * @param k: The repeated times
 * @return: The answer
 */
func longestRepeatingSubsequenceII (str string, k int) int {
    // Write your code here
    n := len(str)
    l, r := 0, n

    for l + 1 < r {
        mid := (l+r)/2
        if getLongest(str, mid) >= k {
            l = mid
        } else {
            r = mid
        }
    }

    if getLongest(str, r) >= k {
        return r
    }
    return l
}

func getLongest(str string, subLen int) int {
    d := make(map[string]int)
    maxCnt := 1
    for i := 0; i < len(str)-subLen+1; i++ {
        d[str[i:i+subLen]]++
        maxCnt = max(maxCnt, d[str[i:i+subLen]])
    }
    return maxCnt
}

func max(a, b int) int {
    if a > b {
        return a
    }
    return b
}

File: lintcode/python/0969_longest_repeating_substring.py

# hash , binary search

from collections import defaultdict

class Solution:
    """
    @param str: The input string
    @param k: The repeated times
    @return: The answer
    """
    def longestRepeatingSubsequenceII(self, input_str, k):
        # Write your code here
        length = len(input_str)
        start, end = 0, length
        while start + 1 < end:
            mid = (start + end) // 2
            if self.count_longest(input_str, mid) >= k:
                start = mid
            else:
                end = mid
        
        # print(start)
        # print(end)
        # print(self.count_longest(input_str, end))
        if self.count_longest(input_str, end) >= k:
            return end
        else:
            return start
    
    
    def count_longest(self, input_str, sub_length):
        cnt = defaultdict(lambda: 0)
        max_cnt = 1
        for i in range(len(input_str) - sub_length + 1):
            cnt[input_str[i: i + sub_length]] += 1
            max_cnt = max(max_cnt, cnt[input_str[i: i + sub_length]])
        return max_cnt
